hermes-frontend/src/main/java/pl/allegro/tech/hermes/frontend/publishing/TimeoutAsyncListener.java

```diff
@@ -26,9 +29,15 @@ public void onComplete(AsyncEvent event) throws IOException {
 
     @Override
     public void onTimeout(AsyncEvent event) throws IOException {
-        if (!messageState.wasDelegatedToKafka()) {
+        requestTimeoutLock.lock();
+        if (messageState.equals(MessageState.State.WAITING_FOR_FIRST_PORTION_OF_DATA) ||
+                messageState.equals(MessageState.State.PARSING) ||
+                messageState.equals(MessageState.State.PARSED)) {
             httpResponder.timeout(event.getThrowable());
+        } else if (messageState.equals(MessageState.State.SENDING_TO_KAFKA)) {
```

Comment:
This else if block is duplicated in BrokerTimeoutAsyncListener but without lock usage. This need to be fixed as we don't want to end of exchange in two places.